### PR TITLE
StyleGAN2-based Generative Model for ADNI Brain Data with t-SNE Visualization

### PR DESCRIPTION
**Pull request by s4698486-Ethan**
This pull request introduces a StyleGAN-based generative model specifically trained on the ADNI brain dataset to generate realistic brain MRI images. It includes:

- The ability to train Generator, Discriminator and Mapping Network models through 150 epochs, generating example brain scans every 2 epochs and saving each of these models every 4 epochs.
- The ability to load saved Generator and Mapping Network models and use them to generate example brain scans, alongside generating style space mappings.

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: StyleGAN on ADNI ---- Hard Difficulty

1. Algorithm solves the problem: (4.5 Marks)
Very clear generated images;
The UMAP should be plotted in colors corresponding to two different classes -0,5

2. Algorithm demonstrated and functions: (3 Marks)

4. Design: (1 Mark)
Highly modularized design

5. Commenting:
Sufficient docstrings but comments could be more specific -0.5 (1 Mark)

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:
- Very detailed documentation, comprehensive justification;
- So the perplexity does make a big difference to the UMAP :). However, I do remember you asked me about the UMAP coloring, but I don't think it's impossible to preserve the labels even the model was trained on AD and NC combined. To do that, you can generate labels randomly (e.g. two labels 0 and 1) when you generate the latent vector `z` (`z_samples = torch.randn(n_samples, z_dim).to(device)`). After the TSNE embedding, use a label to identify a class (e.g. for l in [0, 1]: if labels == l then ...), and plot each class in different color ---- this is just an idea, you can look it up ;
- Good project.

**Comment by s4698486-Ethan** [comment]
Hi Zhixuan, from the above marks it appears that I have scored 19/20 - is this correct? Additionally, would this mean that I would receive 28.5/30 on the overall project (as it is scored out of 30 and I've just scaled 20 up to 30)? 

Thank you for any clarification with this and for the marking!

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: The generated images are realistic. Good explanation of the network. Nice usage guide. Reasonable losses. T-SNE plot titles could be more informative. No clear clusters.
2. **Implementation**: Data augmentation implemented. It is hyperparameters in the constants.py which can be tuned; constants is not a proper name. Overall, nice
3. **Design**: Good
4. **Comment**: Sufficient. Comments are usually under functions
5. **Difficulty**: Hard

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. StyleGAN (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). ON time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-1</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Great work!
